planner/core/logical_plan_builder.go

```diff
@@ -1384,8 +1385,67 @@ func (b *PlanBuilder) buildProjection4Union(ctx context.Context, u *LogicalUnion
 	}
 	u.schema = expression.NewSchema(unionCols...)
 	u.names = names
-	// Process each child and add a projection above original child.
-	// So the schema of `UnionAll` can be the same with its children's.
+
+	allExprs := make([][]expression.Expression, 0)
+	for _, child := range u.children {
+		exprs := make([]expression.Expression, len(child.Schema().Columns))
+		for i, srcCol := range child.Schema().Columns {
+			dstType := unionCols[i].RetType
+			srcType := srcCol.RetType
+
+			if srcType.Tp != dstType.Tp {
+				exprs[i] = castToRes(dstType, u.ctx, srcCol)
+			} else {
+				exprs[i] = srcCol
+			}
+		}
+
+		allExprs = append(allExprs, exprs)
+	}
+
+	// use cast type to modify result charset and collation
+	for i := 0; i < len(u.children[0].Schema().Columns); i++ {
+		tmpExprs := make([]expression.Expression, 0, len(u.Children()))
+		for j := 0; j < len(u.children); j++ {
+			tmpExprs = append(tmpExprs, allExprs[j][i])
+		}
+
+		dstType := unionCols[i].RetType
+		dstType.Charset, dstType.Collate = expression.DeriveCollationFromExprs(b.ctx, tmpExprs...)
+	}
+
+	buildLogicalUnionAll(u, unionCols, b)
+
+	return nil
+}
+
+// castToRes cast to col to result type(ignore result charset collation and collation)
+func castToRes(resType *types.FieldType, ctx sessionctx.Context, col *expression.Column) expression.Expression {
+	switch resType.EvalType() {
+	case types.ETInt:
+		return expression.WrapWithCastAsInt(ctx, col)
+	case types.ETReal:
+		return expression.WrapWithCastAsReal(ctx, col)
+	case types.ETDecimal:
+		return expression.WrapWithCastAsDecimal(ctx, col)
+	case types.ETString:
+		return expression.WrapWithCastAsString(ctx, col)
+	case types.ETDatetime:
+		return expression.WrapWithCastAsTime(ctx, col, types.NewFieldType(mysql.TypeDatetime))
+	case types.ETTimestamp:
+		return expression.WrapWithCastAsTime(ctx, col, types.NewFieldType(mysql.TypeTimestamp))
+	case types.ETDuration:
+		return expression.WrapWithCastAsDuration(ctx, col)
+	case types.ETJson:
+		return expression.WrapWithCastAsJSON(ctx, col)
+	default:
+		return col
+	}
```

Comment:
this part is the same with `newBaseBuiltinFuncWithTp`, maybe we can re-use the code.